feat/add fishtape option to fish feature

edouard-lopez opened the pull request:
## Usage

``json
"features": {
    "ghcr.io/meaningful-ooo/devcontainer-features/fish:1": {
        "fishtape": true
    }
}
``
`


## Options

| Options Id | Description | Type | Default Value |
|-----|-----|-----|-----|
| `fishtape` | Install [Fishtape][fishtape], 100% pure-Fish test runner | boolean | `false` |


- chore: add @devcontainers/cli to devDependencies
- docs: init documentation with install step
- tests: add justfile to run tests and documentation
- feat(fish): add option to install `fishtape`

eitsupi commented:
Thank you for working on this.

But I don't think we should add any more plugins here to be installed by fisher. (I don't think that's a sustainable approach)

Users can install their favorite plug-ins in an instant by simply copying the `~/.config/fish/fish_plugins` file created by fisher into the container and executing the `fisher update` command.

andreiborisov commented:
@edouard-lopez hello, friend! 👋🏻

> But I don't think we should add any more plugins here to be installed by Fisher (I don't think that's a sustainable approach).

Maybe we can use an array for plugins to install instead? @eitsupi @edouard-lopez what do you think?

eitsupi commented:
> Maybe we can use an array for plugins to install instead?

I think adding an array would be a good idea. 👍

eitsupi commented on src/fish/README.md:
```diff
@@ -17,10 +17,12 @@ Installs fish shell and Fisher plugin manager (optionally)
 ---
 
 _Note: This file was auto-generated from the [devcontainer-feature.json](https://github.com/meaningful-ooo/devcontainer-features/blob/main/src/fish/devcontainer-feature.json).  Add additional notes to a `NOTES.md`._
+
+[fisher]: https://github.com/jorgebucaran/fisher
+[fishtape]: https://github.com/jorgebucaran/fishtape
```
Do not edit `README.md` directly as it is automatically generated from `devcontainer-feature.json` and `NOTES.md`.

andreiborisov commented:
> I think adding an array would be a good idea. 👍

The only issue, there is no array type for options: https://containers.dev/implementors/features/#options-property 😓

eitsupi commented:
A string type option that allows specifying plug-ins as a string separated by spaces or commas would be suitable.

```json
"plugins": "jorgebucaran/nvm.fish meaningful-ooo/sponge"
```

andreiborisov commented:
@edouard-lopez is this PR time-sensitive? There is a proposal to add arrays to the spec: https://github.com/devcontainers/spec/issues/57. I prefer to wait for it instead of hardcoding specific plugins.

edouard-lopez commented:
Array sounds good to me, it's not time-sensitive so we can wait or use @eitsupi workaround

andreiborisov commented:
> A string type option that allows specifying plug-ins as a string separated by spaces or commas would be suitable.

Good idea! Let's wait for a reply from the spec maintainers and decide if we want to use this workaround or wait for the proper array support.

edouard-lopez commented:
I updated the code to support a `fisher_plugins` options and added 2 scenarios to test the options.

BTW, Are you aware of solution to test a single scenario with `devcontainer features test`?

andreiborisov commented:
[HAW-29 Option to install Fisher plugins](https://linear.app/meaningful/issue/HAW-29/option-to-install-fisher-plugins)

edouard-lopez commented:
@andreiborisov, are you aware the linear links are not accessible to others?
![image](https://user-images.githubusercontent.com/1212392/200524653-da2a3c52-787f-4d21-8fb5-a4edba9dc764.png)

eitsupi commented on src/fish/devcontainer-feature.json:
```diff
@@ -9,6 +9,11 @@
       "default": true,
       "description": "Install Fisher plugin manager"
+    },
+    "fisher_plugins": {
```
It should be camelCase.

```suggestion
    "fisherPlugins": {
```

eitsupi commented on src/fish/install.sh:
```diff
@@ -118,6 +119,24 @@ if [ "${FISHER}" = "true" ]; then
 fi
 
+# Install Fisher plugins
+if [ -n "${FISHER_PLUGINS}" ]; then
```
fisher should be installed.

```suggestion
if [ "${FISHER}" = "true" ] && [ -n "${FISHERPLUGINS}" ]; then
```

eitsupi commented on src/fish/install.sh:
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env bash
 
 FISHER=${FISHER:-"true"}
+FISHER_PLUGINS=${FISHER_PLUGINS:-""}
```
```suggestion
FISHERPLUGINS=${FISHERPLUGINS:-""}
```

andreiborisov replied:
I propose to go with the same approach maintainers currently use in https://github.com/devcontainers/features: https://github.com/devcontainers/spec/issues/57#issuecomment-1309015728

```bash
BASH_ARRAY=( ${COMMA_SEPARATED_OPTION//,/ } )
```

eitsupi reviewed (changes requested):
Thank you for working on this.
I think the updates regarding the development environment, such as justfile commits, should be separated from the PR, but anyway some small comments regarding the Feature.

eitsupi commented:
> Are you aware of solution to test a single scenario with `devcontainer features test`?

Running only scenario tests can be done as follows, but I believe that running only one scenario is not supported.
We will probably need to rewrite the json file using jq.

https://github.com/meaningful-ooo/devcontainer-features/blob/2177b402ef965dd7a9d5fd5464e8ee7673096b7e/.github/workflows/test.yaml#L70

andreiborisov commented:
> Are you aware the Linear links are not accessible to others?

@edouard-lopez yes, I've wanted to invite you via Twitter (I've lost your Telegram contact 🥲), but it might be that your notifications for DMs are disabled.

andreiborisov reviewed (commented):
Thank you for working on this PR! 🙏🏻

andreiborisov reviewed (changes requested):
We don't have proper commit guidelines yet: https://linear.app/meaningful/issue/SYM-5/commit-style-guide, so I apologize for the chaos.

Although we might arrive at a style similar to what you've used in your commits, right now it's pure gitmoji without scope or type (the type is actually redundant since gitmoji already serves this purpose).

I would love to hear your thoughts on all this if you want to discuss this 🤙🏻

andreiborisov commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,19 @@
+## Install
+
+```console
+yarn install
````
We already use `pnpm`. Essentially there are no reasons not to.

andreiborisov replied:
Contributing guidelines is a huge topic for Meaningful, so it's developed separately organization-wide: https://linear.app/meaningful/team/SYM/active

andreiborisov commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,19 @@
+Pass a `feature` and a `base image`, _e.g._:
+
+```fish
+just test fish mcr.microsoft.com/devcontainers/base:debian
````
There is also https://github.com/go-task/task which looks good. Before committing to something, we should compare those (sorry for being pedantic, I'd like those decisions to be organization-wide so it requires a certain level of scrupulousness).

andreiborisov commented:
@eitsupi thank you for helping with the review, I was quite slow this time 🐢

eitsupi commented:
> > Are you aware of solution to test a single scenario with `devcontainer features test`?
> 
> Running only scenario tests can be done as follows, but I believe that running only one scenario is not supported.

FYI devcontainers/cli#272

eitsupi commented:
Any update on this?
